File: daemon_old/source/offsets.c
#include <stddef.h>
#include <stdint.h>
#include <sys/types.h>
#include <sys/sysctl.h>

#define VERSION_MASK 0xffff0000

#define V300 0x3000000
#define V310 0x3100000
#define V320 0x3200000
#define V321 0x3210000
#define V400 0x4000000
#define V402 0x4020000
#define V403 0x4030000
#define V450 0x4500000
#define V451 0x4510000

static uint32_t get_system_software_version(void) {
	static uint32_t version;
	if (version != 0) {
		return version;
	}
	size_t size = 4;
	sysctlbyname("kern.sdk_version", &version, &size, NULL, 0);
	return version;
}

// NOLINTBEGIN(readability-magic-numbers)

size_t get_allproc_offset(void) {
	static size_t allprocOffset;
	if (allprocOffset != 0) {
		return allprocOffset;
	}
	switch(get_system_software_version() & VERSION_MASK) {
		case V300:
		case V310:
		case V320:
		case V321:
			allprocOffset = 0x276DC58;
			break;
		case V400:
		case V402:
		case V403:
		case V450:
		case V451:
			allprocOffset = 0x27EDCB8;
			break;
		default:
			allprocOffset = -1;
			break;
	}
	return allprocOffset;
}

size_t get_security_flags_offset(void) {
	switch(get_system_software_version() & VERSION_MASK) {
		case V300:
		case V310:
		case V320:
		case V321:
			return 0x6466474;
		case V400:
			return 0x6506474;
		case V402:
		case V403:
		case V450:
		case V451:
			return 0x6505474;
		default:
			return -1;
	}
}

size_t get_qa_flags_offset(void) {
	switch(get_system_software_version() & VERSION_MASK) {
		case V300:
		case V310:
		case V320:
		case V321:
			return 0x6466498;
		case V400:
			return 0x6506498;
		case V402:
			return 0x6505498;
		case V403:
		case V450:
		case V451:
			return 0x6506498;
		default:
			return -1;
	}
}

size_t get_utoken_flags_offset(void) {
	switch(get_system_software_version() & VERSION_MASK) {
		case V300:
		case V310:
		case V320:
		case V321:
			return 0x6466500;
		case V400:
			return 0x6506500;
		case V402:
			return 0x6505500;
		case V403:
		case V450:
		case V451:
			return 0x6506500;
		default:
			return -1;
	}
}

size_t get_root_vnode_offset(void) {
	switch(get_system_software_version() & VERSION_MASK) {
		case V300:
		case V310:
		case V320:
		case V321:
			return 0x67AB4C0;
		case V400:
			return 0x66E74C0;
		case V402:
			return 0x66E64C0;
		case V403:
		case V450:
		case V451:
			return 0x66E74C0;
		default:
			return -1;
	}
}

// NOLINTEND(readability-magic-numbers)

File: daemon/offsets.go
package main

import (
	"fmt"
	"syscall"
)

const (
	VERSION_MASK = 0xffff0000
	V300         = 0x3000000
	V310         = 0x3100000
	V320         = 0x3200000
	V321         = 0x3210000
	V400         = 0x4000000
	V402         = 0x4020000
	V403         = 0x4030000
	V450         = 0x4500000
	V451         = 0x4510000
)

var (
	_version       uint32  = 0
	_allprocOffset uintptr = 0
)

func GetSystemSoftwareVersion() uint32 {
	if _version != 0 {
		return _version
	}
	version, err := syscall.SysctlUint32("kern.sdk_version")
	if err != nil {
		panic("failed to get kernel.sdk_version")
	}
	_version = version
	return _version
}

func GetAllprocOffset() uintptr {
	if _allprocOffset != 0 {
		return _allprocOffset
	}

	version := GetSystemSoftwareVersion() & VERSION_MASK

	switch version {
	case V300:
		fallthrough
	case V310:
		fallthrough
	case V320:
		fallthrough
	case V321:
		_allprocOffset = 0x276DC58
	case V400:
		fallthrough
	case V402:
		fallthrough
	case V403:
		fallthrough
	case V450:
		fallthrough
	case V451:
		_allprocOffset = 0x27EDCB8
	default:
		panic(fmt.Errorf("unsupported kernel version %#08x", version))
	}

	return _allprocOffset
}

func GetSecurityFlagsOffset() uintptr {
	version := GetSystemSoftwareVersion() & VERSION_MASK
	switch version {
	case V300:
		fallthrough
	case V310:
		fallthrough
	case V320:
		fallthrough
	case V321:
		return 0x6466474
	case V400:
		return 0x6506474
	case V402:
		fallthrough
	case V403:
		fallthrough
	case V450:
		fallthrough
	case V451:
		return 0x6505474
	default:
		panic(fmt.Errorf("unsupported kernel version %#08x", version))
	}
}

func GetQaFlagsOffset() uintptr {
	version := GetSystemSoftwareVersion() & VERSION_MASK
	switch version {
	case V300:
		fallthrough
	case V310:
		fallthrough
	case V320:
		fallthrough
	case V321:
		return 0x6466498
	case V400:
		return 0x6506498
	case V402:
		return 0x6505498
	case V403:
		fallthrough
	case V450:
		fallthrough
	case V451:
		return 0x6506498
	default:
		panic(fmt.Errorf("unsupported kernel version %#08x", version))
	}
}

func GetUtokenFlagsOffset() uintptr {
	version := GetSystemSoftwareVersion() & VERSION_MASK
	switch version {
	case V300:
		fallthrough
	case V310:
		fallthrough
	case V320:
		fallthrough
	case V321:
		return 0x6466500
	case V400:
		return 0x6506500
	case V402:
		return 0x6505500
	case V403:
		fallthrough
	case V450:
		fallthrough
	case V451:
		return 0x6506500
	default:
		panic(fmt.Errorf("unsupported kernel version %#08x", version))
	}
}

func GetRootVnodeOffset() uintptr {
	version := GetSystemSoftwareVersion() & VERSION_MASK
	switch version {
	case V300:
		fallthrough
	case V310:
		fallthrough
	case V320:
		fallthrough
	case V321:
		return 0x67AB4C0
	case V400:
		return 0x66E74C0
	case V402:
		return 0x66E64C0
	case V403:
		fallthrough
	case V450:
		fallthrough
	case V451:
		return 0x66E74C0
	default:
		panic(fmt.Errorf("unsupported kernel version %#08x", version))
	}
}
